### PR TITLE
[FIX] mrp: fix report_bom_structure

### PR DESCRIPTION
**Pull request by DavidFesquet**
## Description
When there are 3 levels or more on the BoM structure, the BoM cost is incorrect.
image

## How to reproduce in v.14 to v.15.3 (fixed in v.16):
- Create products M1, M2, M3 and C1.
- Create BoM for 1 unit of M1, using 1 unit of M2
- Create BoM for 1 unit of M2, using 1 unit of M3
- Create BoM for 3 units of M3, using 1 unit of C1
- Set cost of C1 to $ 10.00.
- Go to the BoM of M1 -> 'BoM Structure & Cost' => BoM cost of M1 is $ 1.11, it should be $ 3.33

## Explanation
When there are 3 levels or more on the BoM structure, the 3rd level BoM Cost is divided twice by its quantity (instead of only 1).
It is only when there is 3 level of BoM that _get_price() call itself.
1 level: Computation done in `_get_bom_lines()`
2 levels: Computation done in `_get_bom_lines()` -> `_get_price()`
3 levels: Computation done in `_get_bom_lines()` -> `_get_price()` -> `_get_price()`
The issue is when `get_price()` calls itself, it doesn't send a correct `factor`:
image
The `factor` is already divided by the current `bom.product_qty`, so doing it before recalling `_get_price()` make no sens.

**Result:**
image

**Concerning the Test:**
The original final value was 2.92, it was changed for no reason in this PR: https://github.com/odoo/odoo/pull/79241
I had to put back the original value, else the fix fail the test.

---

UP to saas-15.3

---

OPW-2859407

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by MiquelRForgeFlow — APPROVED** [review]

**Comment by rezak400** [comment]
Iam newbite at github repo mechanism, and i have this problem too in my application, why is this not merged yet on v14 or v15?

**Comment by DavidFesquet** [comment]
Hi @rezak400 , no it is not merged yet, we are still waiting for a review from a senior developer.
Unfortunately, due to the odoo OXP and odoo 16 going out soon, they are currently focusing on urgent matters.

And unfortunately, this bug is not urgent (it doesn't have any impact outside of the report)

**Comment by Whenrow** [comment]
@DavidFesquet hello. i made the same test than yours but with "2" as `product_qty` on each bom and C1 costs 8$. As you can see there is still an issue
image
Your patch work well in case of quantity = 1 only

**Comment by DavidFesquet** [comment]
Hi @Whenrow 
What is the issue with your screenshot?
When you say `"2" as product_qty` I assume that you mean that you created a BoM producing 2 units by using 1 unit of component.

=> C1 is 8 $
* 2 units of M3 = 1 unit of C1
=> M3 is 4 $
* 2 units of M2 = 1 unit of M3
=> M2 is 2 $
* 2 units of M1 = 1 unit of M2
=> M1 is 1 $

The report has the unit cost right.
And the BoM Cost for M1 (first line) is for 2 units of M1, so 2$ is also right.

**Comment by Whenrow** [comment]
@DavidFesquet Hello, your are right I read mine too quickly. Anyway It seems this fix has been merged yesterday (see https://github.com/odoo/odoo/pull/103796)

**Pull request closed without merging**
